VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-detail.js

```diff
@@ -46,6 +46,29 @@ angular.module('virtoCommerce.marketingModule')
             $q.all(blade.expressionPromises).then(function (allData) {
                 blade.currentEntity = angular.copy(data);
                 blade.origEntity = data;
+
```

Comment:
Could you try to get product codes here, instead of all this block:
https://github.com/VirtoCommerce/vc-module-marketing/pull/63/files#diff-ed00182828f55ad18b40aff19684291cR192
There you have `expressionBlock.productIds` array with all ids or `expressionBlock.productId`. In case you have one element in array or `expressionBlock.productId` - need to query `items.query`, create `expressionBlock.productCodes` or `expressionBlock.productCode` based on the result. Then this collection could be used in templates in dynamic expressions module.